### PR TITLE
Fixed interlocking on endpoint deletion [9196]

### PR DESCRIPTION
**Pull request by IkerLuengo**
This PR resolves SOME of the issues with interlocking when deleting endpoints.

Pending issues need further discussion.

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2050/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1807/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2044/)

File: src/cpp/fastdds/subscriber/SubscriberImpl.cpp [hunk 1]
```diff
@@ -210,21 +210,26 @@ ReturnCode_t SubscriberImpl::delete_datareader(
...
     auto it = readers_.find(reader->impl_->get_topicdescription()->get_name());
     if (it != readers_.end())
     {
         auto dr_it = std::find(it->second.begin(), it->second.end(), reader->impl_);
         if (dr_it != it->second.end())
         {
-            (*dr_it)->set_listener(nullptr);
-            (*dr_it)->get_topicdescription()->get_impl()->dereference();
-            delete (*dr_it);
+            //First extract the raeder from the maps to free the mutex
```
Comment (laura-eprosima) [inline src/cpp/fastdds/subscriber/SubscriberImpl.cpp:220]:
```suggestion
            //First extract the reader from the maps to free the mutex
```

File: src/cpp/fastdds/publisher/PublisherImpl.cpp [hunk 2]
```diff
@@ -244,21 +244,26 @@ ReturnCode_t PublisherImpl::delete_datawriter(
...
     auto vit = writers_.find(writer->get_topic()->get_name());
     if (vit != writers_.end())
     {
         auto dw_it = std::find(vit->second.begin(), vit->second.end(), writer->impl_);
         if (dw_it != vit->second.end())
         {
-            (*dw_it)->set_listener(nullptr);
-            (*dw_it)->get_topic()->get_impl()->dereference();
-            delete (*dw_it);
+            //First extract the raeder from the maps to free the mutex
```
Comment (laura-eprosima) [inline src/cpp/fastdds/publisher/PublisherImpl.cpp:254]:
```suggestion
            //First extract the writer from the maps to free the mutex
```

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2067/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1823/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2061/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2072/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1828/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2066/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2075/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1831/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2069/)

File: src/cpp/rtps/builtin/discovery/participant/PDP.cpp [hunk 3]
```diff
@@ -695,6 +695,8 @@ ReaderProxyData* PDP::addReaderProxyData(
                 RTPSParticipantListener* listener = mp_RTPSParticipant->getListener();
                 if (listener)
                 {
+                    //Unlock before calling listeners
+                    lockPDP.unlock();
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/participant/PDP.cpp:699]: I know this was the main purpose of this PR, but we cannot unlock here, as `ReaderDiscoveryInfo` just holds a const ref to `ReaderProxyData` and not a copy.

The same applies to `WriterDiscoveryInfo` and `WriterProxyData`.

So I am afraid we should revert all changes on this file.

Comment (IkerLuengo) [reply src/cpp/rtps/builtin/discovery/participant/PDP.cpp]: So true.
reverted.

**Review by MiguelCompany — DISMISSED** [review]

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2082/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1838/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2076/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2518/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2331/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2525/)

**Review by laura-eprosima — DISMISSED** [review]

**Review by laura-eprosima — APPROVED** [review]

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2603/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2419/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2613/)

**Comment by MiguelCompany** [comment]
Failures seem unrelated. Going in

**Review by MiguelCompany — APPROVED** [review]
LGTM

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2888/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/188/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2643/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2899/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3018/)
* Linux aarch64Build Status
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2705/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3029/)

**Comment by MiguelCompany** [comment]
@IkerLuengo This one needs a rebase

**Comment by IkerLuengo** [comment]
Rebased to 2.0.x/HEAD

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3075/)
* Linux aarch64Build Status
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2753/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3093/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3370/)
* Linux aarch64Build Status
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3019/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3402/)

**Comment by JLBuenoLopez-eProsima** [comment]
This PR is not going to make it to `v2.0.3` which is going to be the last release because `2.0.x` is becoming EOL (end of life). Thus, I am going to proceed and close this PR.

**Pull request closed without merging**